File: C++/ex25/huffman_coding_solution.cpp
/**
 * Exercise 25: Huffman Coding (Data Compression)
 * Difficulty: ***** (5/5)
 * 
 * Implement Huffman coding to compress text data.
 * Build a Huffman tree from character frequencies and generate codes.
 * 
 * Input:
 * - A string to be encoded
 * 
 * Output:
 * - Print frequency of each character
 * - Print Huffman code for each character
 * - Print total bits needed for encoding
 */

#include <bits/stdc++.h>
using namespace std;

struct Node {
    char ch;
    int freq;
    Node* left;
    Node* right;
    
    Node(char c, int f) : ch(c), freq(f), left(nullptr), right(nullptr) {}
};

struct Compare {
    bool operator()(Node* a, Node* b) {
        return a->freq > b->freq;
    }
};

void generateCodes(Node* root, string code, unordered_map<char, string>& codes) {
    if (!root) return;
    
    if (!root->left && !root->right) {
        codes[root->ch] = code;
        return;
    }
    
    if (root->left) {
        generateCodes(root->left, code + "0", codes);
    }
    
    if (root->right) {
        generateCodes(root->right, code + "1", codes);
    }
}

int main() {
    string text;
    getline(cin, text);
    
    unordered_map<char, int> freq;
    for (char c : text) {
        freq[c]++;
    }
    
    priority_queue<Node*, vector<Node*>, Compare> pq;
    
    for (auto& p : freq) {
        pq.push(new Node(p.first, p.second));
    }
    
    while (pq.size() > 1) {
        Node* left = pq.top();
        pq.pop();
        Node* right = pq.top();
        pq.pop();
        
        Node* parent = new Node('\0', left->freq + right->freq);
        parent->left = left;
        parent->right = right;
        
        pq.push(parent);
    }
    
    Node* root = pq.top();
    pq.pop();
    
    unordered_map<char, string> codes;
    if (root->left || root->right) {
        generateCodes(root, "", codes);
    } else {
        codes[root->ch] = "0";
    }
    
    long long totalBits = 0;
    for (char c : text) {
        totalBits += codes[c].length();
    }
    
    for (auto& p : freq) {
        cout << p.first << ": " << p.second << " bits = " << codes[p.first] << "\n";
    }
    
    cout << "Total bits: " << totalBits << "\n";
    
    return 0;
}

File: C#/ex25/huffman_coding_solution.cs
/**
 * Exercise 25: Huffman Coding (Data Compression)
 * Difficulty: ***** (5/5)
 * 
 * Implement Huffman coding to compress text data.
 * Build a Huffman tree from character frequencies and generate codes.
 * 
 * Input:
 * - A string to be encoded
 * 
 * Output:
 * - Print frequency of each character
 * - Print Huffman code for each character
 * - Print total bits needed for encoding
 */

using System;
using System.Collections.Generic;
using System.Linq;

class Node {
    public char Ch { get; set; }
    public int Freq { get; set; }
    public Node Left { get; set; }
    public Node Right { get; set; }
    
    public Node(char c, int f) {
        Ch = c;
        Freq = f;
        Left = null;
        Right = null;
    }
}

class NodeComparer : IComparer<Node> {
    public int Compare(Node a, Node b) {
        return a.Freq.CompareTo(b.Freq);
    }
}

class Program {
    static void GenerateCodes(Node root, string code, Dictionary<char, string> codes) {
        if (root == null) return;
        
        if (root.Left == null && root.Right == null) {
            codes[root.Ch] = code;
            return;
        }
        
        if (root.Left != null) {
            GenerateCodes(root.Left, code + "0", codes);
        }
        
        if (root.Right != null) {
            GenerateCodes(root.Right, code + "1", codes);
        }
    }
    
    static void Main() {
        string text = Console.ReadLine();
        
        Dictionary<char, int> freq = new Dictionary<char, int>();
        foreach (char c in text) {
            if (freq.ContainsKey(c)) {
                freq[c]++;
            } else {
                freq[c] = 1;
            }
        }
        
        var pq = new SortedSet<Node>(new NodeComparer());
        
        foreach (var p in freq) {
            pq.Add(new Node(p.Key, p.Value));
        }
        
        while (pq.Count > 1) {
            Node left = pq.Min;
            pq.Remove(left);
            Node right = pq.Min;
            pq.Remove(right);
            
            Node parent = new Node('\0', left.Freq + right.Freq);
            parent.Left = left;
            parent.Right = right;
            
            pq.Add(parent);
        }
        
        Node root = pq.Min;
        
        Dictionary<char, string> codes = new Dictionary<char, string>();
        if (root.Left != null || root.Right != null) {
            GenerateCodes(root, "", codes);
        } else {
            codes[root.Ch] = "0";
        }
        
        long totalBits = 0;
        foreach (char c in text) {
            totalBits += codes[c].Length;
        }
        
        foreach (var p in freq) {
            Console.WriteLine($"{p.Key}: {p.Value} bits = {codes[p.Key]}");
        }
        
        Console.WriteLine($"Total bits: {totalBits}");
    }
}
